runner/runner.go

```diff
@@ -0,0 +1,379 @@
+	workersRunEvents int
+
+	heartbeatInt int
+
+	disabledLoopCount    uint64
+	eventRunErrCount     uint64
+	eventRunSuccessCount uint64
+
+	logger  *log.Logger
+	logDest string
+	debug   bool
+)
+
+const getEventsLoop time.Duration = time.Minute
+const getEventsBreak time.Duration = time.Second
+const runEventsBreak time.Duration = time.Second * 10
+
+func init() {
+	flag.StringVar(&wpCliPath, "cli", "/usr/local/bin/wp", "Path to WP-CLI binary")
+	flag.IntVar(&wpNetwork, "network", 0, "WordPress network ID, `0` to disable")
+	flag.StringVar(&wpPath, "wp", "/var/www/html", "Path to WordPress installation")
+	flag.IntVar(&workersGetEvents, "workers-get", 1, "Number of workers to retrieve events")
+	flag.IntVar(&workersRunEvents, "workers-run", 5, "Number of workers to run events")
+	flag.IntVar(&heartbeatInt, "heartbeat", 60, "Heartbeat interval in seconds")
+	flag.StringVar(&logDest, "log", "os.Stdout", "Log path, omit to log to Stdout")
+	flag.BoolVar(&debug, "debug", false, "Include additional log data for debugging")
+	flag.Parse()
+
+	setUpLogger()
+
+	// TODO: Should check for wp-config.php instead?
+	validatePath(&wpCliPath)
+	validatePath(&wpPath)
+}
+
+func main() {
+	logger.Printf("Starting with %d event-retreival worker(s) and %d event worker(s)", workersGetEvents, workersRunEvents)
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
+
+	sites := make(chan site)
+	events := make(chan event)
+
+	go spawnEventRetrievers(sites, events)
+	go spawnEventWorkers(events)
+	go retrieveSitesPeriodically(sites)
+	go heartbeat()
+
+	caughtSig := <-sig
+	logger.Printf("Stopping, got signal %s", caughtSig)
+}
+
+func spawnEventRetrievers(sites <-chan site, queue chan<- event) {
+	for w := 1; w <= workersGetEvents; w++ {
+		go queueSiteEvents(w, sites, queue)
+	}
+}
+
+func spawnEventWorkers(queue <-chan event) {
+	workerEvents := make(chan event)
+
+	for w := 1; w <= workersRunEvents; w++ {
+		go runEvents(w, workerEvents)
+	}
+
+	for event := range queue {
+		workerEvents <- event
+	}
+
+	close(workerEvents)
+}
+
+func retrieveSitesPeriodically(sites chan<- site) {
+	for {
+		siteList, err := getSites()
+		if err != nil {
+			time.Sleep(getEventsLoop)
+			continue
+		}
+
+		for _, site := range siteList {
+			sites <- site
+		}
+
+		time.Sleep(getEventsLoop)
+	}
+}
+
+func heartbeat() {
+	if heartbeatInt == 0 {
+		logger.Println("heartbeat disabled")
+		return
+	}
+
+	interval := time.Duration(heartbeatInt) * time.Second
+	time.Sleep(interval)
+
+	for {
+		successCount, errCount := atomic.LoadUint64(&eventRunSuccessCount), atomic.LoadUint64(&eventRunErrCount)
+		atomic.SwapUint64(&eventRunSuccessCount, 0)
+		atomic.SwapUint64(&eventRunErrCount, 0)
+		logger.Printf("<heartbeat eventsSucceededSinceLast=%d eventsErroredSinceLast=%d>", successCount, errCount)
+
+		time.Sleep(interval)
+	}
+}
+
+func getSites() ([]site, error) {
+	siteInfo, err := getInstanceInfo()
+	if err != nil {
+		return make([]site, 0), err
```

Comment:
Only if the function that threw the error returns the same data type, which isn't the case here. `getInstanceInfo()` returns a `siteInfo` struct. I'll check the other cases, but I believe that I've only used `make` where a different data type is called for.